File: day18.c
#include <stdio.h>
#define W 71
#define H 71

int G[H][W]={0};
int Q[1000]={0};
int read=0,write=0;
void resetQ(void){read=0;write=0;}
void enq(int x){Q[write]=x; write=(write+1)%1000;}
int deq(void){int x=Q[read]; read=(read+1)%1000; return x;}
int empty(void){return read==write;}

int main(void){
    FILE *f=fopen("data/input18.txt","r");
    int nobst=0,x=0,y=0;
    int Lobst[5000][2]={0};
    while(fscanf(f,"%d,%d",&x,&y)==2){
        G[y][x]=nobst+1; Lobst[nobst][0]=x; Lobst[nobst][1]=y;
        nobst++;
    }
    int solveFor(int k){
        int seen[H][W]={0};
        int sx=0,sy=0,ex=W-1,ey=H-1;
        resetQ();enq(0);enq(0);enq(0);
        while(!empty()){
            int d=deq(),x=deq(),y=deq();
            if(x==ex && y==ey)return d;
            if(seen[y][x])continue;
            seen[y][x]=1;
            if(x>0 && !(G[y][x-1]>0 && G[y][x-1]-1<=k)){enq(d+1);enq(x-1);enq(y);}
            if(x<W-1 && !(G[y][x+1]>0 && G[y][x+1]-1<=k)){enq(d+1);enq(x+1);enq(y);}
            if(y>0 && !(G[y-1][x]>0 && G[y-1][x]-1<=k)){enq(d+1);enq(x);enq(y-1);}
            if(y<H-1 && !(G[y+1][x]>0 && G[y+1][x]-1<=k)){enq(d+1);enq(x);enq(y+1);}
        }
        return 999999;
    }
    int part1=solveFor(1024);
    int left=1024,right=nobst;
    while(right-left>1){
        int middle=(left+right)/2;
        if(solveFor(middle)==999999){right=middle;}else{left=middle;}
    }
    printf("%d %d,%d\n",part1,Lobst[right][0],Lobst[right][1]);
    return 0;
}

File: day18.lua
io.input("data/input18.txt")
local inp=io.read("*all")
local obst={}
local Lobst={}
function hash(x,y) return 100*y+x end
local nobst=0
for x,y in string.gmatch(inp,"(%d+),(%d+)") do
    obst[hash(x,y)]=nobst; table.insert(Lobst,{x,y}); nobst=nobst+1
end

function solveFor(k)
    local W,H=71,71
    local sx,sy,ex,ey=0,0,W-1,H-1
    local Q,seen={{0,sx,sy}},{}
    while #Q>0 do
        local u=table.remove(Q,1)
        local d,x,y=u[1],u[2],u[3]
        if x==ex and y==ey then return d end
        if not seen[hash(x,y)] then
            seen[hash(x,y)]=true
            for _,a in ipairs({{x-1,y},{x+1,y},{x,y-1},{x,y+1}}) do         
                local xx,yy=a[1],a[2]
                local h=hash(xx,yy)
                if xx>=0 and xx<W and yy>=0 and yy<H
                and not(obst[h] and obst[h]<=k) and not seen[h] then
                    table.insert(Q,{d+1,xx,yy})
                end
            end
        end
    end
end

local part1=solveFor(1024)
local left,right=1024,nobst
while right-left>1 do
    local middle=(left+right)//2
    if solveFor(middle) then left=middle else right=middle end
end
local a=Lobst[right+1]
local part2=a[1]..","..a[2]
print(part1,part2)